Major refactor

Jfeatherstone opened the pull request:
Thank you for creating and working on this package!

I apologize for submitting such a large pull request; ideally I would have liked to break it up into several smaller ones, but I had quite some trouble understanding different parts of the code, and the only way I could work my changes into the current codebase without it being hacky was by refactoring large sections. Below are some of the reasons why I found it difficult to make my changes:

- All of the code is in a single file, which doesn't always make it clear which parts actually relate to which other ones.

- Many parts of the code, whether in the functions used, the code style, or the logic expressions, make it clear that this package was written for Python 2.x, and then adapted for Python 3.x. There are also certain practices that make it difficult to easily understand what keyword arguments do or what a function returns. One example: it was hard to understand several functions such as `powerlaw.pdf`, `powerlaw.cdf`, etc. because of the very liberal use of `**kwargs`.

- There aren't many comments throughout the code, and the documentation is not very deep for most functions or variables. Especially since this package is based on several papers, I think it is important to be able to quickly determine where the methods or expressions are coming from.

- Given the fact that there are several distributions implemented with underlying similarities, it is understandable that chunks of code are reused throughout the project. That being said, I had quite some difficulty trying to change something (limits on power law exponent) since I had to track several different sections that all looked similar and somehow all needed to be changed to get the desired result.

- I think the fitting and comparing methods in this library are excellent, and become more valuable as more distributions are implemented. As such, I think there is quite some value to making sure that distributions are easy and straight-forward to implement. I think the current system is not the easiest.

With that in mind, this pull request includes a major refactor of pretty much the entire package, with much more documentation, as well as several bug fixes and improvements. The front-end user experience should be almost exactly the same as before, with the exception of a few points discussed below (bug fixes are listed after).

- The API documentation was pretty lacking, and I think a manuscript as the documentation for a package that is constantly being updated is not a good idea. I've updated the API documentation for pretty much all functions, and rewritten the sphinx documentation to include all the information from the manuscript, as well as more tutorials about implementing new distributions, why this library differs from others, etc. For now, you see the documentation hosted here: http://jfeatherstone.xyz/powerlaw_docs_temp/.

- The reference datasets are now included in the package itself, and can be easily loaded for testing with `powerlaw.load_test_dataset()`.

- You now access parameters of a distribution by directly using their name (eg. `dist.alpha`, `fit.exponential.Lambda`) instead of using `dist.parameter1`; I think this is much more clear, and scalable. I've added a debug function that still assigns the properties `parameter1`, etc. for backwards compatibility, but I guess this should be removed eventually. 

- As per #113, I have changed the alpha default range from `[1, 3]` to `[0, 3]`. On top of EliasL's reasoning in the linked Issue, it is also possible that you might want to fit a distribution that genuinely has a power law exponent between 0 and 1, not just close to 1. While these distributions are non-normalizable for an infinite domain, there are many examples of real distributions that scale with an exponent less than 1 within a finite range (see Clauset et al. 2009, Section 3 for a discussion of this). EliasL submitted a separate pull request for this, but this only partially fixes the issue. See #113 for more in depth discussion, but this pull request changes all of the necessary elements for this to work.

- I've added many more checks throughout the code to help with debugging, including adding several warnings using the `warnings` module. This includes things like if you fit a parameter value that is close to the edge of it's allowed range (suggested by EliasL, see #113), if the `estimate_discrete` keyword is used in a situation where it might lead to errors (suggested by henrymartin1, see #104) if a valid fit isn't found, if you don't have enough data to calculate certain quantities, if there might be numerical errors based on the parameters values, etc. Some of these were already there, and have now been converted to be proper warnings with `warnings.warn`. I've also created a [tutorial](http://jfeatherstone.xyz/powerlaw_docs_temp/tutorials/warnings.html) on the documentation site about what each of these warnings mean and how you might address them if you encounter any particular one.

- The keywords for `Fit` about discrete approximations are quite confusing, since there is `discrete_approximation` which is general and works for all distributions and controls computing the PDF, but then specific distributions like `Power_Law` have `estimate_discrete` which is also a discrete approximation but for the parameters. These sound like they should do the same thing, but in fact they address quite different aspects of the fitting/generation.\
I think it is better to rename `discrete_approximation` to `discrete_normalization` since indeed this parameter controls how to approximate the normalization of the PDF. Then I think it is more clear to call the two options `'round'` and `'sum'`, where `'sum'` replaces `'xmax'`. This would remove the option to pass an integer to use as the maximum value which you sum up to, but I don't actually see a use case for that. Please let me know if anyone does, but if you want to specify an upper limit, you should do it for the whole fitting process, not just for the normalization.\
`estimate_discrete` can stay how it is, since it would now not be easily confused with `discrete_normalization`.

- Similarly, I've removed the `pdf_ends_at_xmax` keyword. The user specifying a value for `xmax` should be taken as signal that the `xmax` is an upper bound for the PDF, so it is confusing to have a second keyword argument 
that isn't obviously different. The only thing this affects is the normalizer for power laws, for which it makes more sense just to see if `xmax` has a value. Again, please let me know if I'm missing a use case for this, but nothing is mentioned in the documentation, and I can't think of one.

- I have switched the main numerical fitting function (`Distribution.fit()`) to use `scipy.optimize.minimize()` instead of `scipy.optimize.fmin()`. This former function is more extensible and allows more control of the fitting process. For example, in the past, parameter bounds were only used to tell if the fitting worked; using `minimize()` the bounds are actually used in the fitting process. This also allows the use of both bounds (fixed parameter ranges) and constraints (relational checks between parameters) at the same time; previously these were both called "ranges", and only one could be used at a time. For more info, please check out the tutorial on "Parameter ranges and constraints". There is a lot to say about choosing the arguments of `minimize()` and I've tried to put as much detail in comments around the relevant sections, but I expect there is still some fine-tuning to do here. But for now, the fits perform at least as well as they used to. I've included a small section in the new documentation that covers some of the changes and nuances in usage (under Tutorials -> Advanced Topics).

- For progress bars, I've replaced the custom implementations with `tqdm`, a pretty standard and very flexible progress bar library.

- I've added a custom implementation of `__dir__` for the Fit class so that you can autocomplete the supported distributions even before they actually exist as properties.

- It is inconvenient that the word/variable "sigma" can refer to two different things in the code: in one case, it is the standard error of the MLE (eg. for the power law, see Clauset et al. 2009, Eq. 3.2), and in the other, it is a parameter for a distribution (lognormal width). I have changed all references that have the latter meaning to "width", and added comments noting where the expressions are sourced from so it still can be linked to the original works despite the difference in name. I guess technically they have similar meanings, but the way they are used in the code is very different, and you wouldn't want to confuse, for example, parameter ranges for one with the other.

This pull request fixes the following bugs:

- As per #114, there was a logic issue with the `Distribution.parent_Fit` variable that was causing errors when you pass both a value for `xmin` and `parameter_ranges` to `powerlaw.Fit`. I've fixed this by making sure that `parent_Fit` is certainly defined, and the logic checks only if it's value is `None` or not, as opposed to seeing if the property exists.

- There was a bug in the `Exponential` class where if you have a large value for `xmin` ( > 1000 or so), computing the normalizer can cause an overflow error since it is calculated as `exp(lambda * xmin)`. Eventually this is multiplied by the pdf which will have something like `exp(-lambda * x)`, and since x is going to be at least as large as xmin, their product will be a reasonable value. But storing the two variables separately can lead to needing to store values that can be as big as 1e1000. It's not a perfect fix, but for now I've just stored these values as numpy.float128 types. An ideal solution might use an infinite precision library like mpmath or decimal, but those come with quite some computational slowdowns, so if this works as is, I think it's the best solution.

- `powerlaw.pdf` and `powerlaw.plot_pdf` didn't properly account for having zero or negative values in the passed data, unlike `powerlaw.Distribution.pdf`. This means that you could end up with `log(0)` errors when plotting the pdf without the `Fit` class. An example dataset that has zeros is `quakes.txt`.

- `powerlaw.plot_*` functions raised an error if you tried to plot a distribution that didn't have data, ie. if you created the distribution by assigning a fixed value of the parameters. For example, the following used to raise an error:

```
pl = powerlaw.Power_Law(xmin=1, parameters=[2.0])
pl.plot_pdf()
```

This now works as you would expect it to.

----

Ok that's the end of the changes!

Just to make sure I haven't messed up anything, I haven't touched the testing files at all, and all of the tests still pass. I did have to put a debug function to assign some variable names that I changed; see `Distribution.debug_parameter_names()`.

I've bumped the version up to "1.6.0" in both the source file and the `setup.py` file.

Please let me know if you have any comments or suggestions.

Jfeatherstone commented:
Sure, I'll work on some updates to the tests in another PR.

Also, I think the workflow tests should all pass now. Sorry it took a few tries; I forgot to test with multiple versions of python and on different platforms :)

keflavich commented on .gitignore:
```diff
@@ -0,0 +1,196 @@
+## Images
```
This file has a lot of irrelevant file types.   Could you filter it down to only appropriate ones for a python package?

keflavich replied:
There are still a lot of irrelevant types here

Jfeatherstone commented:
The most recent commits have included changes to the random number generation, which I've realized had a few bugs. It took me quite a while to debug it, and was only able to identify the issue after adding more tests (see #116). With these changes, all tests in #116 pass.

Comments:
- The expression for the inverse CDF for lognormal distributions was in fact the inverse _CCDF_ (I believe; the important part is that it wasn't the inverse CDF). For unbounded random number generation, it doesn't matter which one you use, but if you want to put upper or lower limits on the random numbers, every distribution in the package needs to consistently use either the CDF or the CCDF. I've switched this to the inverse CDF. For future reference (and I think it's a good tutorial in case someone wants to learn about inverse transform sampling, since I couldn't find many others online) I've explicitly derived the transforms for every distribution in the (temporarily-hosted) documentation [here](https://jfeatherstone.xyz/powerlaw_docs_temp/derivations_top.html). Along with some changes to the bounds in `Distribution.generate_random`, all distributions should generate random numbers that obey xmin and xmax (if it exists). \
 Another reason that I wanted to post these derivations is because the stretched exponential defined in the library is not the same as that used in Clauset et al. (2009): the difference is in whether the exponential argument is `lambda x^beta` (Clauset et al. version) or `(lambda x)^beta` (current powerlaw version). I think the latter is more physical because the former would lead to lambda having weird units, but no comment is made about this in the package, so just linking to that paper as the source of the analytics makes things confusing. Having derivations for the explicit form used in the package avoids this confusion. \
 Finally --- and I plan on opening an issue so we can discuss this in its own thread --- I've just realized that what is called "stretched exponential" in this library and in Clauset et al. (2009) is actually the Weibull distribution. The stretched exponential doesn't have the `x^(beta - 1)` factor in front of the exponential. So far I haven't changed anything, but it would be a good idea to rename this (and add the true stretched exponential distribution).
- I've added many "TODO" flags on things that either aren't clear to me, or that I know need to be fixed in a future update. I intend to work on most of these myself, though of course if others want to weigh in, please do.

keflavich commented on powerlaw/distributions.py:
```diff
@@ -0,0 +1,2964 @@
+        # (but we don't define them here because we don't want to overwrite
+        # the values created in the subclass)
+        # And the name of the distribution
+        #name = 'name'
```
these might be definable using ABCMeta methods, though I'm not super familiar with those so I'm not requesting a change, just noting the possibility.

keflavich commented on powerlaw/distributions.py:
```diff
@@ -0,0 +1,2964 @@
+        It is not currently implemented since the value will depend on your
+        specific distribution, so it should be implemented in child classes.
+        """
+        return False
```
```suggestion
        raise NotImplementedError
```

keflavich commented on powerlaw/distributions.py:
```diff
@@ -0,0 +1,2964 @@
+        possible_numerical_error = False
+        from numpy import isnan, min
+        if isnan(min(CDF)):
+            print("'nan' in fit cumulative distribution values.", file=sys.stderr)
```
this should be a Warning instead of a print-to-stderr

keflavich commented:
This is big enough to warrant a major version number change, so my plan is to merge this and tag it v2.0.  I'll integrate some infrastructure upgrades too.

@Jfeatherstone given how much work you've put into this, and the todo items you have listed in the code, I'm opening a [Discussion](https://github.com/jeffalstott/powerlaw/discussions/117) about how to proceed with this.  I think you ought to be brought on as a maintainer with commit rights.

Jfeatherstone commented on powerlaw/distributions.py:
```diff
@@ -0,0 +1,2964 @@
+        # TODO: I would have thought that rounding properly here would give
+        # more accurate distributions, but, specifically for discrete truncated power
+        # law distributions, this actually leads to very bad fits. It could
+        # have something to do with the fact
```
I realized as I was writing that comment that I actually have no idea what is causing that... :) I guess I just forgot to delete that last sentence.

My initial thought was something to do with a numerical error given that the discrete truncated power law uses some pretty heavy special functions (namely `gamminc` and `lerchphi` from `mpmath`). But that doesn't really explain why we should consistently round values down (as casting to int does).

When I get around to investigating this, a good start would be to figure out how the rounding affects other distributions; the only test that actually fails with proper rounding is truncated power law, but it's possible that the other distributions also get less accurate, but still accurate enough to pass tests. Things like this are why I created the `accuracy_overview.py` script in #116.

Jfeatherstone commented:
Excellent, thanks for the feedback, and for creating the discussion post!

Since you've incorporated things into the v2.0 branch, would you prefer I include the requested changes in a new pull request there? Or maybe just wait until I have commit rights and directly add my changes, since the last few things are quite small? Or I can continue to update this PR.

I would say the only discussion that I would like to further bring up here relates to the name of the parameters for the lognormal distribution. In the original PR, I renamed lognormal's `sigma` parameter to `width` to avoid conflict with the standard error of the power law, also called `sigma`. After working with the code for a couple months in making the subsequent changes, I think this was the wrong choice: the lognormal parameter should remain named `sigma`, and the standard error of the power law distribution should be instead changed to something like `standard_error` or `s_err`. Especially given how constraints are structured differently from parameter ranges now, this way makes more sense to me. If you have any thoughts on this, please let me know.
